### PR TITLE
CORDA-995: r3corda wire compatibility

### PR DESCRIPTION
**Pull request by exFalso**
This PR introduces wire compatibility between `corda` and `r3corda`. Hopefully. We haven't tested it yet with proper nodes, opening now to parallelise review and testing as we're time constrained for 3.0

The PR includes adjustments in SMM/FSMI, removal of TopicSession and removal of the exposed send ack handler in MessagingService.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt [hunk 1]
```diff
@@ -1,59 +1,122 @@
...
+import net.corda.core.flows.FlowInfo
 import net.corda.core.serialization.CordaSerializable
-import net.corda.core.serialization.SerializationDefaults
 import net.corda.core.serialization.SerializedBytes
-import net.corda.core.utilities.UntrustworthyData
-import java.io.IOException
+import java.security.SecureRandom
 
+/**
+ * A session between two flows is identified by two session IDs, the initiating and the initiated session ID.
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: This comment is a little off

**Comment by mnesbit** [comment]
I want confirmation that we have manually run an actual interop test and that an automated test is coming. I can't tell from this if we have actually run bank of corda demo for instance with one node being open source and confirmed that the cash is transferrable back and forth.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/InMemoryMessagingNetwork.kt [hunk 2]
```diff
@@ -438,8 +438,8 @@ class InMemoryMessagingNetwork internal constructor(
         }
 
         private fun MessageTransfer.toReceivedMessage(): ReceivedMessage = InMemoryReceivedMessage(
-                message.topicSession,
-                message.data.copyOf(), // Kryo messes with the buffer so give each client a unique copy
+                message.topic,
+                OpaqueBytes(message.data.bytes.copyOf()), // Kryo messes with the buffer so give each client a unique copy
```
Comment (mnesbit) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/InMemoryMessagingNetwork.kt]: Can this be removed now we use AMQP?

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt [hunk 3]
```diff
@@ -354,11 +354,11 @@ class StateMachineManagerImpl(
             val session = FlowSessionInternal(
                     flow,
                     flowSession,
-                    random63BitValue(),
+                    SessionId.createRandom(newSecureRandom()),
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt]: This looks very expensive

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt [hunk 4]
```diff
@@ -336,12 +336,12 @@ class StateMachineManagerImpl(
...
     }
 
-    private fun onSessionInit(sessionInit: SessionInit, receivedMessage: ReceivedMessage, sender: Party) {
+    private fun onSessionInit(sessionInit: InitialSessionMessage, receivedMessage: ReceivedMessage, sender: Party) {
 
         logger.trace { "Received $sessionInit from $sender" }
         val senderSessionId = sessionInit.initiatorSessionId
 
-        fun sendSessionReject(message: String) = sendSessionMessage(sender, SessionReject(senderSessionId, message))
+        fun sendSessionReject(message: String) = sendSessionMessage(sender, ExistingSessionMessage(senderSessionId, RejectSessionMessage(message, random63BitValue())))
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt]: The random63BitValue looks inconsistent with the sessionid.createrandom below?

File: node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt [hunk 5]
```diff
@@ -1,59 +1,122 @@
...
+ *
+ *             Flow A (initiating)                   Flow B (initiated)
+ *           initiatingId=sourceId=0
+ *        send(Initiate(initiatingId=0)) ----->    initiatingId=sinkId=0
+ *                                                 initiatedId=sourceId=1
+ *            initiatedId=sinkId=1       <----- send(Confirm(initiatedId=1))
+ */
 @CordaSerializable
-interface SessionMessage
+sealed class SessionMessage
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: Is this going to give us extensibility issues later? e.g. if we require a diffie-hellman step for E2E encryption?

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowIORequest.kt [hunk 6]
```diff
@@ -70,7 +70,7 @@ class ReceiveAll(val requests: List<ReceiveRequest<SessionData>>) : WaitingReque
             if (isComplete(receivedMessages)) {
                 receivedMessages
             } else {
-                throw IllegalStateException(requests.filter { it.session !in receivedMessages.keys }.map { "Was expecting a ${it.receiveType.simpleName} but instead got nothing for $it." }.joinToString { "\n" })
+                throw IllegalStateException(requests.filter { it.session !in receivedMessages.keys }.map { "Was expecting a message but instead got nothing for $it." }.joinToString { "\n" })
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowIORequest.kt]: Knowing the receive type is useful for debugging.

Comment (exFalso) [reply node/src/main/kotlin/net/corda/node/services/statemachine/FlowIORequest.kt]: This part of the code will be replaced soon, which is why I tried to do this part quick-and-dirty

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt (hunk 3 above)
Comment (exFalso) [reply node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt]: I'll just leave this here
```kotlin
fun random63BitValue(): Long {
    while (true) {
        val candidate = Math.abs(newSecureRandom().nextLong())
        // No need to check for -0L
        if (candidate != 0L && candidate != Long.MIN_VALUE) {
            return candidate
        }
    }
}
```
haha

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt (hunk 4 above)
Comment (exFalso) [reply node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt]: This isn't a sessionID

File: node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt (hunk 5 above)
Comment (exFalso) [reply node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: `sealed`-ness doesn't appear in bytecode, so in theory we can still extend this. However even with that I don't think the protocol is forwards compatible

**Comment by exFalso** [comment]
@mnesbit wrt testing that's currently underway

**Comment by exFalso** [comment]
@mnesbit test complete, merging this

File: node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt (hunk 5 above)
Comment (mikehearn) [reply node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: It kind of does - the c'tor becomes private, right?

**Pull request merged**